algorithms.go

```diff
@@ -93,37 +111,43 @@ func tokenBucket(s Store, c Cache, r *RateLimitReq) (resp *RateLimitResp, err er
 					return nil, err
 				}
 			}
-			// If our new duration means we are currently expired
-			if expire < MillisecondNow() {
-				// Update this so s.OnChange() will get the new expire change
-				item.ExpireAt = expire
-				c.Remove(item.Key)
-				return tokenBucket(s, c, r)
+
+			// If our new duration means we are currently expired.
+			now := MillisecondNow()
+			if expire <= now {
+				// Renew item.
+				expire = now + r.Duration
+				t.CreatedAt = now
+				t.Remaining = t.Limit
```

Comment:
This doesn't cover the case where the duration changed because we switched to `Behavior_DURATION_IS_GREGORIAN`. Could we call `tokenBucketNewItem()` here to cover that case?

Reply:
We don't have to check for expiration with `Behavior_DURATION_IS_GREGORIAN` because it is always a future time.